local_working_copy: when all sides of a conflict are executable, mate…

gulbanana opened the pull request:
local_working_copy: when all sides of a conflict are executable, materialise the conflicted file as executable

Fixes #3579 and adds a testcase for an executable conflict treevalue.

gulbanana commented:
The new test fails before the fix and passes after it. This repro script can also be used to verify:

```sh
#!/usr/bin/env bash

mkdir repro_permfail
cd repro_permfail/

jj git init

cat <<EOF > script.sh
#!/usr/bin/bash
echo "No conflict here."
EOF
chmod 755 script.sh

jj desc -m "script import"
jj new -m "set up for conflict"
jj branch create setup-conflict
cat <<EOF >> script.sh
echo "Half a conflict."
EOF

jj new -r @- -m "create a conflict"

cat <<EOF >> script.sh
echo "The other half."
EOF

jj rebase -d setup-conflict
```

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -1238,8 +1239,7 @@ impl TreeState {
             })?;
         let size = conflict_data.len() as u64;
-        // TODO: Set the executable bit correctly (when possible) and preserve that on
-        // Windows like we do with the executable bit for regular files.
```
I think the Windows issue still exist. The comment can be moved to `write_path_to_store()`?

martinvonz replied:
Agreed. And it's probably not very hard to fix it there if you're interested, but feel free to leave it for later.

gulbanana replied:
I see, the filemode issue doesn't repro on Windows so I didn't realise there was another one. I'll look deeper.

gulbanana replied:
ok, i think i understand this now - windows filesystems may not have `+x`, so we keep the last known value of the executable flag for ordinary files. i've done the same thing for conflicts:
```rs
                    #[cfg(windows)]
                    let executable = {
                        let () = executable; // use the variable
                        if let Some(merge) = current_tree_values.to_executable_merge() {
                            merge.into_resolved().unwrap_or(Some(false)).unwrap_or_default()
                        } else {
                            false
                        }
                    };
```

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -1223,6 +1223,7 @@ impl TreeState {
         &self,
         disk_path: &Path,
         conflict_data: Vec<u8>,
+        executable: bool,
```
copy to `FileState::for_file(_, ..)`?

gulbanana replied:
right, i missed the hardcoded `false` there

yuja commented on lib/src/merge.rs:
```diff
@@ -501,6 +501,13 @@ impl MergedTreeValue {
+        self.values
+            .iter()
+            .all(|v| matches!(*v, Some(TreeValue::File { executable, .. }) if executable))
+    }
```
handle `is_present()` and `None` like `.is_tree()`?

Absent values should be excluded as we do in `to_file_merge()`. I'm not sure if an empty merge should be considered an executable, though.

gulbanana replied:
Alright, I'll change that (or maybe look for a version with slightly less unwrap ;_;).

gulbanana replied:
ok, i've got it using a merge of the executable bits now, and we only consider the conflict treevalue executable if that merge resolves

gulbanana commented:
Thanks for the reviews. I've addressed the feedback, and added a second commit which fixes some clippy warnings which were output only on windows - turns out that the `cfg`ed out behaviour around the executable bit has been triggering lints, but our PR checks weren't catching them.

martinvonz commented:
> added a second commit which fixes some clippy warnings which were output only on windows - turns out that the `cfg`ed out behaviour around the executable bit has been triggering lints, but our PR checks weren't catching them.

Thanks. I suppose we could also run clippy on windows. Not sure it's worth the resource cost given how infrequently it's going to find anything.

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -1392,8 +1397,13 @@ impl TreeState {
-                MaterializedTreeValue::Conflict { id: _, contents } => {
-                    self.write_conflict(&disk_path, contents)?
+                MaterializedTreeValue::Conflict { id, contents } => {
+                    let executable = if let Some(merge) = id.to_executable_merge() {
+                        merge.resolve_trivial().copied().unwrap_or_default()
+                    } else {
+                        false
+                    };
```
I think it would make sense to make the materialized executable bit part of `MaterializedTreeValue::Conflict`. That would also make the "mode" field in `jj diff --git` a bit more accurate. Feel free to leave that for later.

martinvonz reviewed (approved):
Thanks!

gulbanana commented on CHANGELOG.md:
```diff
@@ -16,7 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed bugs
-
+* Files with conflicts are now checked out as executable if all sides of the 
+  conflict are executable.
```
The merge-based calculation is more sophisticated, but I think it still has basically the same effect. It's only if all sides of the merge are
- present
- executable 
that it gets trivially resolved to `true`. 

Or is there some other case that makes this not an [iff](https://en.wikipedia.org/wiki/If_and_only_if)?
